File: src/converter/plugins/InjectionPlugin.ts
import ts, {Node} from "typescript";
import {ConverterContext} from "../context.js";
import {Injection, InjectionContext, InjectionType} from "../injection.js";
import {ConverterPlugin} from "../plugin.js";
import {GeneratedFile} from "../generated.js";
import {Render} from "../render.js";

export const injectionServiceKey = Symbol()

export class InjectionService {
    constructor(readonly injections: Injection[]) {
    }

    resolveInjections(
        node: Node,
        type: InjectionType,
        context: ConverterContext,
        render: Render,
    ): string[] {
        const injectionContext: InjectionContext = {
            ...context,
            static: type === InjectionType.STATIC_MEMBER,
            type,
        }

        return this.internalResolveInjections(node, injectionContext, render)
    }

    private internalResolveInjections(
        node: Node,
        context: InjectionContext,
        render: Render
    ) {
        const injections: string[] = []

        for (const injection of this.injections) {
            const result = injection.inject(node, context, render)

            if (result !== null) injections.push(result)
        }

        return injections
    }
}

export class InjectionPlugin implements ConverterPlugin {
    private readonly injectionService: InjectionService

    constructor(injections: Injection[]) {
        this.injectionService = new InjectionService(injections)
    }

    generate(): GeneratedFile[] {
        return [];
    }

    render(node: Node, context: ConverterContext, next: Render): string | null {
        return null;
    }

    setup(context: ConverterContext): void {
        context.registerService(injectionServiceKey, this.injectionService)
    }

    traverse(node: ts.Node, context: ConverterContext): void {
    }
}

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/InjectionPlugin.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.*
import js.array.ReadonlyArray
import js.symbol.Symbol
import typescript.Node

@OptIn(ExperimentalJsExport::class)
@JsExport
val injectionServiceKey = Symbol()

class InjectionService(private val injections: ReadonlyArray<Injection<Node, Node>>) {
    fun resolveInjections(
        node: Node,
        type: InjectionType,
        context: ConverterContext,
        render: Render<Node>,
    ): ReadonlyArray<String> {
        val injectionContext = object : InjectionContext, ConverterContext by context {
            override val type = type
        }

        return internalResolveInjections(node, injectionContext, render)
    }

    private fun internalResolveInjections(
        node: Node,
        context: InjectionContext,
        render: Render<Node>
    ): ReadonlyArray<String> {
        val injections = mutableListOf<String>()

        for (injection in this.injections) {
            val result = injection.inject(node, context, render)

            if (result != null) injections += result
        }

        return injections.toTypedArray()
    }
}

class InjectionPlugin(injections: ReadonlyArray<Injection<Node, Node>>) : ConverterPlugin<Node> {
    private val injectionService = InjectionService(injections)

    override fun traverse(node: Node, context: ConverterContext) = Unit

    override fun render(node: Node, context: ConverterContext, next: Render<Node>) = null

    override fun generate(context: ConverterContext, render: Render<Node>) = emptyArray<GeneratedFile>()

    override fun setup(context: ConverterContext) {
        context.registerService(injectionServiceKey, this.injectionService)
    }
}
